server/odc-service/src/main/java/com/oceanbase/odc/service/permission/database/DatabasePermissionHelper.java

```diff
@@ -0,0 +1,153 @@
+ */
+package com.oceanbase.odc.service.permission.database;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.apache.commons.collections4.CollectionUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import com.oceanbase.odc.core.shared.constant.ResourceRoleName;
+import com.oceanbase.odc.core.shared.exception.AccessDeniedException;
+import com.oceanbase.odc.metadb.connection.DatabaseEntity;
+import com.oceanbase.odc.metadb.connection.DatabaseRepository;
+import com.oceanbase.odc.metadb.iam.UserDatabasePermissionEntity;
+import com.oceanbase.odc.metadb.iam.UserDatabasePermissionRepository;
+import com.oceanbase.odc.service.collaboration.project.ProjectService;
+import com.oceanbase.odc.service.iam.auth.AuthenticationFacade;
+import com.oceanbase.odc.service.permission.database.model.DatabasePermissionType;
+
+/**
+ * @author gaoda.xy
+ * @date 2024/1/15 10:58
+ */
+@Component
+public class DatabasePermissionHelper {
+
+    @Autowired
+    private ProjectService projectService;
+
+    @Autowired
+    private AuthenticationFacade authenticationFacade;
+
+    @Autowired
+    private DatabaseRepository databaseRepository;
+
+    @Autowired
+    private UserDatabasePermissionRepository userDatabasePermissionRepository;
+
+    /**
+     * Check whether the current user has the permission to access the database
+     * 
+     * @param databaseIds database ids
+     * @param permissionTypes permission types
+     */
+    public void checkPermissions(Collection<Long> databaseIds, Collection<DatabasePermissionType> permissionTypes) {
+        if (CollectionUtils.isEmpty(databaseIds) || CollectionUtils.isEmpty(permissionTypes)) {
+            return;
+        }
+        List<DatabaseEntity> entities = databaseRepository.findByIdIn(databaseIds);
+        List<Long> toCheckDatabaseIds = new ArrayList<>();
+        Set<Long> projectIds = getPermittedProjectIds();
+        for (DatabaseEntity e : entities) {
+            if (e.getProjectId() == null) {
+                throw new AccessDeniedException("Database is not belong to any project");
+            }
+            if (!projectIds.contains(e.getProjectId())) {
+                toCheckDatabaseIds.add(e.getId());
+            }
+        }
+        Map<Long, Set<DatabasePermissionType>> id2PermissionTypes = getDatabaseId2PermissionTypes(databaseIds);
+        for (Long databaseId : toCheckDatabaseIds) {
+            Set<DatabasePermissionType> authorized = id2PermissionTypes.get(databaseId);
+            if (CollectionUtils.isEmpty(authorized)) {
+                throw new AccessDeniedException(String.format("No permission for the database with id %s", databaseId));
+            }
+            Set<DatabasePermissionType> unAuthorized =
```

Comment:
What should we do if there is an implicit relationship between permissions? (Of course there is none now)

Reply:
For now, let's keep it that way, for the sake of O(N) time complexity. If we use `Permission#implies()`, the complexity is O(N^2).